vm/src/compiler.rs

```diff
@@ -226,38 +213,40 @@ impl CompilerEnv for TypeInfos {
     }
 }
 
-pub struct Compiler<'a> {
-    globals: &'a (CompilerEnv + 'a),
+pub struct Compiler<'a, E> {
```

Comment:
`Compiler` (just as `Typecheck`) are one of the places where two monomorphizations can actually mean a signficant amount of duplicated code. I think its better to keep the trait object here for that reason